etc/proto/protoc-gen-zap/main.go

```diff
@@ -0,0 +1,302 @@
+				g.P("enc.AddObject(\"", fname, "\", x.", gname, ")")
+			} else {
+				fmt.Fprintf(os.Stderr, "** FALLBACK ON %v\n", f.Desc.Message().FullName())
+
+				g.P("if obj, ok := interface{}(x.", gname, ").(", g.QualifiedGoIdent(zapcorePkg.Ident("ObjectMarshaler")), "); ok {")
+				g.P("enc.AddObject(\"", fname, "\", obj)")
+				g.P("} else {")
+				g.P("enc.AddReflected(\"", fname, "\", x.", gname, ")")
+				g.P("}")
+			}
+		}
+	case protoreflect.StringKind:
+		if half {
+			g.P(g.QualifiedGoIdent(extensionPkg.Ident("AddHalfString")), `(enc, "`, fname, `", x.`, gname, `)`)
+		} else {
+			g.P("enc.AddString(\"", fname, "\", x.", gname, ")")
+		}
+	default:
+		g.P("enc.AddReflected(\"", fname, "\", x.", gname, ")")
+	}
+}
+
+func isPachydermProto(fullName string) bool {
+	if strings.Contains(fullName, "_v2.") {
+		return true
+	}
+	return strings.HasPrefix(fullName, "datum.") ||
+		strings.HasPrefix(fullName, "pfsload.") ||
+		strings.HasPrefix(fullName, "taskapi.")
+}
+
+func generateMessage(g *protogen.GeneratedFile, m *protogen.Message) {
+	ident := g.QualifiedGoIdent(m.GoIdent)
+	g.P("func (x *", ident, ") MarshalLogObject(enc ", g.QualifiedGoIdent(zapcorePkg.Ident("ObjectEncoder")), ") error {")
+	g.P("if x == nil {")
+	g.P("return nil")
+	g.P("}")
+	for _, f := range m.Fields {
+		opts := f.Desc.Options().(*descriptorpb.FieldOptions)
+		if opts != nil {
+			if customName := proto.GetExtension(opts, gogoproto.E_Customname).(string); customName != "" {
+				f.GoName = customName
+			}
+		}
+		if f.Desc.Name() == "size" && f.Desc.ContainingMessage().FullName() == "pfs_v2.Trigger" {
+			// For some reason, this field gets generated as Size_ instead of Size.
+			// It's because there is code in the package that implements a Size method,
+			// but it's unclear to me how the proto compiler can know this.
+			f.GoName = "Size_"
+		}
+		if proto.GetExtension(opts, protoextensions.E_Mask).(bool) {
+			g.P("enc.AddString(\"", f.Desc.Name(), "\", \"[MASKED]\")")
+		} else if f.Desc.IsList() {
+			generateListField(g, f)
+		} else if f.Desc.IsMap() {
+			generateMapField(g, f)
+		} else {
+			generatePrimitiveField(g, f, opts)
+		}
+	}
+	g.P("return nil")
+	g.P("}")
+	g.P()
+	for _, submsg := range m.Messages {
+		if submsg.Desc.IsMapEntry() {
+			continue
+		}
+		generateMessage(g, submsg)
+	}
+}
+
+func generateFile(gen *protogen.Plugin, file *protogen.File) *protogen.GeneratedFile {
+	if len(file.Messages) == 0 {
+		return nil
+	}
+
+	filename := fmt.Sprintf("%s.pb.zap.go", file.GeneratedFilenamePrefix)
+	g := gen.NewGeneratedFile(filename, file.GoImportPath)
+	g.P("// Code generated by protoc-gen-zap (etc/proto/protoc-gen-zap). DO NOT EDIT.")
+	g.P("//")
+	g.P("// source: ", file.Desc.Path())
+	g.P()
+	g.P("package ", file.GoPackageName)
+	g.P()
+
+	for _, m := range file.Messages {
+		generateMessage(g, m)
+	}
+
+	return g
+}
+
+func main() {
+	protogen.Options{}.Run(func(plugin *protogen.Plugin) error {
```

Comment:
This is a neat way to do the entrypoint.

Reply:
Copied and pasted from the godoc 😂